core/src/main/java/me/mykindos/betterpvp/core/combat/events/DamageEvent.java

```diff
@@ -1,150 +1,433 @@
-        this.damage += dam;
+    public Multimap<ModifierType, DamageModifier> getModifiers() {
+        return ArrayListMultimap.create(modifiers);
     }
 
     /**
-     * Takes a certain amount of damage away from the already existing damage amount
-     *
-     * @param dam The amount of damage to remove
+     * Gets all modifiers of a specific type applied to this event
+     * @param type the modifier type to filter by
+     * @return list of modifiers of the specified type
      */
-    public void removeDamage(double dam) {
-        this.damage -= dam;
+    public List<DamageModifier> getModifiers(ModifierType type) {
+        return List.copyOf(modifiers.get(type));
+    }
+
+    public List<DamageModifier> getAppliedModifiers() {
+        final ArrayList<DamageModifier> modifiers = new ArrayList<>(this.modifiers.values());
+        final ArrayList<DamageModifier> resultList = new ArrayList<>();
+        modifiers.sort(Comparator.comparingInt(DamageModifier::getPriority).reversed());
+
+        for (DamageModifier modifier : modifiers) {
+            if (isModifierExcluded(modifier) || !modifier.canApply(this)) {
+                continue;
+            }
+
+            ModifierResult result = modifier.apply(this);
+            Preconditions.checkNotNull(result, "Modifier %s returned null result", modifier.getName());
+
+            // Apply the modification
+            resultList.add(modifier);
+
+            // If this modifier cancels others, stop processing
+            if (result.isCancelOtherModifiers()) {
+                break;
+            }
+        }
+
+        return resultList;
     }
 
     /**
-     * Gets the damage modifiers for this event.
-     * Listeners can use this to add modifiers that will be processed at the end of the event.
-     *
-     * @return The damage modifiers
+     * Calculates the final modified damage after applying all modifiers in order of priority
+     * @return the final modified damage
      */
-    public DamageModifiers getDamageModifiers() {
-        return damageModifiers;
+    public double getModifiedDamage() {
+        final ArrayList<DamageModifier> modifiers = new ArrayList<>(this.modifiers.values());
+        modifiers.sort(Comparator.comparingInt(DamageModifier::getPriority).reversed());
+
+        double damage = getDamage();
+        for (DamageModifier modifier : modifiers) {
+            if (isModifierExcluded(modifier) || !modifier.canApply(this)) {
+                continue;
+            }
+
+            ModifierResult result = modifier.apply(this);
+            Preconditions.checkNotNull(result, "Modifier %s returned null result", modifier.getName());
+
+            // Apply the modification
+            damage = applyModifierResult(damage, result);
+
+            // If this modifier cancels others, stop processing
+            if (result.isCancelOtherModifiers()) {
+                break;
+            }
+        }
+
+        return damage;
     }
```

Comment:
by priority is going to end up being a mess. I think we need some sort of consistent behaviour e.g. damage increases are processed first, and damage decreases are processed last